webapp/channels/src/components/dynamic_virtualized_list/item_row.tsx

```diff
@@ -0,0 +1,98 @@
+// See LICENSE.txt for license information.
+
+import debounce from 'lodash/debounce';
+import type {ReactNode} from 'react';
+import React, {memo, useEffect, useRef} from 'react';
+
+const RESIZE_DEBOUNCE_TIME = 120; // in ms
+
+interface Props {
+    item: ReactNode;
+    itemId: string;
+    index: number;
+    height: number;
+    width?: number; // Its optional since it may not be available when the parent component is mounted
+    onHeightChange: (itemId: string, height: number, forceScrollCorrection: boolean) => void;
+    onUnmount: (itemId: string, index: number) => void;
+}
+
+/**
+ * This component is used to measure the height of a row and update the height of the row when it changes.
+ */
+const ItemRow = (props: Props) => {
+    const rowRef = useRef<HTMLDivElement>(null);
+
+    const itemIdRef = useRef(props.itemId);
+    const heightRef = useRef(props.height);
+    const widthRef = useRef(props.width);
+    const indexRef = useRef(props.index);
+
+    // This prevents stale closures in the ResizeObserver callback
+    // and ensures event handlers always have the latest values
+    // we also update these refs whenever their source values change
+    useEffect(() => {
+        heightRef.current = props.height;
+        widthRef.current = props.width;
+        itemIdRef.current = props.itemId;
+        indexRef.current = props.index;
+    }, [props.itemId, props.height, props.width, props.index]);
+
+    // This effect is used to measure the height of the row as soon as the component mounts
+    useEffect(() => {
+        const newHeight = Math.ceil(rowRef?.current?.offsetHeight ?? 0);
+        props.onHeightChange(itemIdRef.current, newHeight, false);
+    }, []);
+
+    useEffect(() => {
+        const debouncedOnHeightChange = debounce((height: number) => {
+            // If width of container has changed then scroll bar position will be out of sync
+            // so we need to force a scroll correction
+            const forceScrollCorrection = rowRef.current?.offsetWidth !== widthRef.current;
+
+            heightRef.current = height;
+
+            props.onHeightChange(itemIdRef.current, height, forceScrollCorrection);
+        }, RESIZE_DEBOUNCE_TIME);
+
+        function resizeObserverCallback(resizeEntries: ResizeObserverEntry[]) {
+            if (!rowRef.current) {
+                return;
+            }
+
+            // Since we're observing a single row, we can safely assume that the first entry is the one we want
+            if (resizeEntries.length === 1 && resizeEntries[0].target === rowRef.current) {
+                const newHeight = Math.ceil(resizeEntries[0].borderBoxSize[0].blockSize);
+
+                // If the height has changed significantly, update the height
+                if (newHeight !== heightRef.current) {
+                    debouncedOnHeightChange(newHeight);
+                }
+            }
+        }
+
+        const resizeObserver = new ResizeObserver(resizeObserverCallback);
```

Comment:
Correct me if I am wrong, but this will create a new `ResizeObserver` per item row right? This could potentially lead to thousands of resizeObservers depending on how many posts a user has.

Given that a resizeObserver can observe multiple elements, would it make sense to have a global singleton observer, and then observe/unobserve elements per item row? That could make it even more efficient on memory.

You would only need to change the callback logic where you assume that the first entry is the one you want.

Reply:
I am performing tests currently to see how much of benifit shared resizable observer

Reply:
It turns out there was a slight benifit with shared resize observer in terms of memory, so i have updated it